### PR TITLE
Enable keyword substitution in emails on enrollment

### PR DESCRIPTION
**Pull request by kluo**
In studio, an instructor can include keywords in their course enrollment email which will be automatically
substituted with the corresponding value for the recipient.
The keywords supported are:

%%USER_ID%% => anonymous_user_id
%%USER_FULLNAME%% => user profile name
%%COURSE_DISPLAY_NAME%% => display name of the course
%%COURSE_END_DATE%% => end date of the course

This uses the same validation code as the existing keyword validator for the LMS, but works with CMS which uses requirejs.

@jbau @stvstnfrd

### CODE REVIEW

**Comment by stvstnfrd** [comment]
@kluo Is the enrollment email feature exclusive to our fork?

File: cms/static/coffee/src/keyword_validator.coffee [hunk 1]
```diff
@@ -0,0 +1,28 @@
+define ->
```
Comment (stvstnfrd) [inline cms/static/coffee/src/keyword_validator.coffee]: New code/files should not be written in Coffeescript, which has been deprecated in our codebase.

File: cms/startup.py [hunk 2]
```diff
@@ -21,6 +23,11 @@ def run():
 
     add_mimetypes()
 
+    # Monkey patch the keyword function map
```
Comment (stvstnfrd) [inline cms/startup.py]: Can you explain what's going on here?

_Why_ are we monkey patching the function map?

File: cms/djangoapps/contentstore/views/course.py [hunk 3]
```diff
@@ -728,15 +729,16 @@ def course_info_update_handler(request, course_key_string, provided_id=None):
...
     """
     Handles ajax request for sending test enrollment emails to the instructor
     """
+    course_key = CourseKey.from_string(course_key_string)
     user = request.user
     from_address = microsite.get_value('email_from_address', settings.DEFAULT_FROM_EMAIL)
     subject = request.POST.get('subject')
     subject = ''.join(subject.splitlines())
-    message = request.POST.get('message')
+    message = substitute_keywords_with_data(request.POST.get('message'), user.id, course_key)
```
Comment (stvstnfrd) [inline cms/djangoapps/contentstore/views/course.py]: I tend to prefer making this kind of change like the following:
```python
message = request.POST.get('message')
message = substitute_keywords_with_data(message, user.id, course_key)
```

- The diff is cleaner, as it's just an added line, not both an added and
  removed line.
- It's easier to debug, as you can step through lines of code and see
  how the variable changes from call to call.
- It's easier to read, as nested function calls often lead to longer
  lines which are more difficult to mentally parse with levels of nested
  parens.

File: cms/static/js/views/settings/main.js [hunk 4]
```diff
@@ -367,12 +367,19 @@ var DetailsView = ValidatingView.extend({
             message = this.post_enrollment_email_elem.val();
         }
 
-        $.post("/settings/send_test_enrollment_email",
+        // Validation for keyword substitution
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: Nit: This comment seems superfluous ("The `KeywordValidator` is for 'Validation for keyword...'").

File: cms/static/js/views/settings/main.js [hunk 5]
```diff
@@ -367,12 +367,19 @@ var DetailsView = ValidatingView.extend({
...
+        var validation = KeywordValidator.validate_string(message);
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: Please hoist this variable and declare it at the top of the function.

File: cms/static/js/views/settings/main.js [hunk 6]
```diff
@@ -367,12 +367,19 @@ var DetailsView = ValidatingView.extend({
...
+        if (!validation.is_valid) {
+            message = gettext("There are invalid keywords in your email. Please check the following keywords and try again:");
+            message += "\n" + validation.invalid_keywords.join('\n');
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: We can include the newline at the end of the string literal above:
```javascript
message = gettext('There are invalid keywords in your email. Please check the following keywords and try again:\n');
message += validation.invalid_keywords.join('\n');
```

File: cms/static/js/views/settings/main.js [hunk 7]
```diff
@@ -367,12 +367,19 @@ var DetailsView = ValidatingView.extend({
...
+            alert(message);
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: We should explicitly call `window.alert(...)` instead of relying on this in the global namespace.

File: cms/static/js/views/settings/main.js [hunk 8]
```diff
@@ -367,12 +367,19 @@ var DetailsView = ValidatingView.extend({
             message = this.post_enrollment_email_elem.val();
         }
 
-        $.post("/settings/send_test_enrollment_email",
+        // Validation for keyword substitution
+        var validation = KeywordValidator.validate_string(message);
+        if (!validation.is_valid) {
+            message = gettext("There are invalid keywords in your email. Please check the following keywords and try again:");
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: Use single quotes here for consistency.

File: cms/static/js/views/settings/main.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
 define(["js/views/validation", "codemirror", "underscore", "jquery", "jquery.ui", "js/utils/date_utils", "js/models/uploads",
-    "js/views/uploads", "js/utils/change_on_enter", "jquery.timepicker", "date"],
-    function(ValidatingView, CodeMirror, _, $, ui, DateUtils, FileUploadModel, FileUploadDialog, TriggerChangeEventOnEnter) {
+    "js/views/uploads", "js/utils/change_on_enter", "coffee/src/keyword_validator", "jquery.timepicker", "date"],
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: `jquery.timepicker` and `data` don't appear to be used; they aren't passed as params to the function on the following line.

But maybe this is outside the scope and should be a separate PR to upstream?

**Comment by stvstnfrd** [comment]
A few minor points and requests for clarification.

The biggest thing I'd like to see here is the (new) Coffeescript file rewritten in vanilla JS, as CS has been deprecated from our codebase.

Where is this feature located within CMS such that I can test it manually/locally.

**Comment by kluo** [comment]
Currently yes

File: cms/djangoapps/contentstore/views/course.py (hunk 3 above)
Comment (kluo) [inline cms/djangoapps/contentstore/views/course.py]: All good points, I'll fix that

File: cms/startup.py (hunk 2 above)
Comment (kluo) [inline cms/startup.py]: https://github.com/Stanford-Online/edx-platform/blob/master/lms/startup.py#L161

File: cms/static/js/views/settings/main.js (hunk 9 above)
Comment (kluo) [inline cms/static/js/views/settings/main.js]: No they are used, I think they work as globals

File: cms/static/js/views/settings/main.js (hunk 4 above)
Comment (kluo) [inline cms/static/js/views/settings/main.js]: True

File: cms/static/js/views/settings/main.js (hunk 8 above)
Comment (kluo) [inline cms/static/js/views/settings/main.js]: There are still tons of other places using double quotes, should there be a separate PR to tidy those up as well?

File: cms/static/js/views/settings/main.js (hunk 7 above)
Comment (kluo) [inline cms/static/js/views/settings/main.js]: This is also used everywhere else there is an alert, should that be a separate PR to make everything explicit? Although I assume it's understood not to redefine alert as another function, so hypothetically it should be fine. But I get what you are saying, better safe than sorry.

Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: It could be, though for now I'd be content to just see it changed here (in new code).

There's also a performance consideration with implicit global variables; the interpreter has to walk the entire scope chain on each reference to resolve globals.

File: cms/static/js/views/settings/main.js (hunk 8 above)
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: It could be, though for now I'd be content to just see it changed here (in new code) and in any other lines you've changed for this PR.

File: cms/static/coffee/src/keyword_validator.coffee (hunk 1 above)
Comment (jbau) [inline cms/static/coffee/src/keyword_validator.coffee]: @kluo is this picked over from NickDup's old code?

Comment (kluo) [inline cms/static/coffee/src/keyword_validator.coffee]: It's the same code as used on the LMS side, but since I've got his vanilla javascript for the CMS checked out now I'm not using this anymore

Comment (jbau) [inline cms/static/coffee/src/keyword_validator.coffee]: ok.  please remove this when possible then

**Comment by jbau** [comment]
I just finished a pass.  looks ok to me, module @stvstnfrd's comments.

**Comment by jbau** [comment]
probably want a prod demo at some point

File: cms/static/js/views/settings/main.js [hunk 10]
```diff
@@ -367,12 +368,18 @@ var DetailsView = ValidatingView.extend({
             message = this.post_enrollment_email_elem.val();
         }
 
-        $.post("/settings/send_test_enrollment_email",
+        validation = ViewUtils.keywordValidator.validate_string(message);
+        if (!validation.is_valid) {
+            message = gettext('There are invalid keywords in your email. Please check the following keywords and try again:\n');
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: My bad... I know I asked you to pull the `\n` up here, but looking at it now, I realize we shouldn't be passing a newline (or any formatting markers) on to translation functions.

Can you revert this and add the `\n` in the line below? Sorry about that.

File: cms/static/js/views/settings/main.js [hunk 11]
```diff
@@ -367,12 +368,18 @@ var DetailsView = ValidatingView.extend({
...
+            message += validation.invalid_keywords.join('\n');
+            window.alert(message);
+            return;
+        }
+
+        $.post($(event.target).data('endpoint'),
                {
                  subject: subject,
-                 message:message,org:this.model.get('org'),
-                 number: this.model.get('course_id'),
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: Why don't we need these vars anymore? What were they used for?

File: cms/static/js/views/utils/view_utils.js [hunk 12]
```diff
@@ -186,6 +214,7 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
             'setScrollOffset': setScrollOffset,
             'redirect': redirect,
             'reload': reload,
-            'hasChangedAttributes': hasChangedAttributes
+            'hasChangedAttributes': hasChangedAttributes,
+            'keywordValidator': keywordValidator,
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Sadly, JavaScript only (relatively) recently added support for trailing commas in object/array definitions; this causes syntax errors in certain browsers/runtimes.

We'll get a clean diff if you move `keywordValidator` above `hasChangedAttributes` and remove the trailing comma from the `hasChangedAttributes` line.

File: cms/static/js/views/utils/view_utils.js [hunk 13]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
             return false;
         };
 
+        keywordValidator = (function () {
+            var regexp = /%%+[^%]+%%/g;
+            var keywords = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_END_DATE%%'];
+            var validate = function (string) {
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: This should be a named function.

Named functions are defined at compile time, whereas anonymous functions assigned to variables are defined at runtime, among other subtleties [1].

An unsung benefit to named functions over anonymous functions is that they provide a more explicit stack trace when debugging.

```javascript
function validate(string) {
    // ...
}
```

[1] http://stackoverflow.com/questions/336859/var-functionname-function-vs-function-functionname

File: cms/static/js/views/utils/view_utils.js [hunk 14]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
+                var regex_match = string.match(regexp);
+                var found_keywords = regex_match == null ? [] : regex_match;
+                var invalid_keywords = [];
+                var num_found = found_keywords.length;
+                var curr_keyword;
+
+                $.each(found_keywords, function(i, keyword) {
+                    if ($.inArray(keyword, keywords) == -1) {
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Use the strict equality operator `===`.

File: cms/static/js/views/utils/view_utils.js [hunk 15]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
+                        invalid_keywords.push(keyword);
+                    }
+                });
+
+                return {
+                    'is_valid': invalid_keywords.length == 0,
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Use the strict equality operator `===`.

Or you could write `!invalid_kewords.length`, though Jason will disagree :P (I need to remember to put together a compelling case/response for this convention in dynamic languages)

File: cms/static/js/views/utils/view_utils.js [hunk 16]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
             return false;
         };
 
+        keywordValidator = (function () {
+            var regexp = /%%+[^%]+%%/g;
+            var keywords = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_END_DATE%%'];
+            var validate = function (string) {
+                var regex_match = string.match(regexp);
+                var found_keywords = regex_match == null ? [] : regex_match;
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Simplify this as:
```javascript
var foundKeywords = regexMatch || [];
```

File: cms/static/js/views/utils/view_utils.js [hunk 17]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
+                var invalid_keywords = [];
+                var num_found = found_keywords.length;
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: I don't think we're ever using the `num_found` variable and it can be deleted.

File: cms/static/js/views/utils/view_utils.js [hunk 18]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
+                });
+
+                return {
+                    'is_valid': invalid_keywords.length == 0,
+                    'invalid_keywords': invalid_keywords,
+                }
+
+            };
+            return {
+                'validate_string': validate,
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Remove trailing comma (see note elsewhere in thread).

File: cms/static/js/views/utils/view_utils.js [hunk 19]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
+
+                $.each(found_keywords, function(i, keyword) {
+                    if ($.inArray(keyword, keywords) == -1) {
+                        invalid_keywords.push(keyword);
+                    }
+                });
+
+                return {
+                    'is_valid': invalid_keywords.length == 0,
+                    'invalid_keywords': invalid_keywords,
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Remove trailing comma (see note elsewhere in thread).

File: cms/static/js/views/utils/view_utils.js [hunk 20]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
 
+        keywordValidator = (function () {
+            var regexp = /%%+[^%]+%%/g;
+            var keywords = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_END_DATE%%'];
+            var validate = function (string) {
+                var regex_match = string.match(regexp);
+                var found_keywords = regex_match == null ? [] : regex_match;
+                var invalid_keywords = [];
+                var num_found = found_keywords.length;
+                var curr_keyword;
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: I don't think we're ever using the `curr_keyword` variable and it can be deleted.

File: cms/static/js/views/utils/view_utils.js [hunk 21]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
+
+                $.each(found_keywords, function(i, keyword) {
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: It feels like `$.map` would be a more natural fit here [1].

```javascript
var keywordsInvalid = $.map(keywordsFound, function (keyword) {
    if ($.inArray(keyword, keywordsSupported) === -1) {
        return keyword;
    } else {
        // return `null` or `undefined` to remove an element
        return undefined;
    }
});
```

[1] http://api.jquery.com/jquery.map/

File: cms/static/js/views/utils/view_utils.js [hunk 22]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
             return false;
         };
 
+        keywordValidator = (function () {
+            var regexp = /%%+[^%]+%%/g;
+            var keywords = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_END_DATE%%'];
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: This should be named more specifically to differentiate it from the
other arrays of keywords, e.g. `keywordsSupported`.

Prefixing each of the keyword arrays like `keywordsSupported`,
`keywordsInvalid`, and `keywordsFound` make the variables easier to grok
at a glance.

I'm a big of some form of Hungarian notation; Spolsky provides an
interesting read, _Making Wrong Code Look Wrong_ [1].

[1] http://www.joelonsoftware.com/articles/Wrong.html

File: cms/static/js/views/utils/view_utils.js [hunk 23]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
             return false;
         };
 
+        keywordValidator = (function () {
+            var regexp = /%%+[^%]+%%/g;
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Why are we matching 2 _or more_ leading percents but only _exactly_ 2 trailing percents?

File: cms/static/js/views/utils/view_utils.js [hunk 24]
```diff
@@ -5,7 +5,8 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
     function ($, _, gettext, NotificationView, PromptView) {
         var toggleExpandCollapse, showLoadingIndicator, hideLoadingIndicator, confirmThenRunOperation,
             runOperationShowingMessage, disableElementWhileRunning, getScrollOffset, setScrollOffset,
-            setScrollTop, redirect, reload, hasChangedAttributes, deleteNotificationHandler;
+            setScrollTop, redirect, reload, hasChangedAttributes, deleteNotificationHandler,
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Declaring each variable on a separate line presents a cleaner diff.  At a glance, it looks like this entire line is changing, whereas we're just adding a new declaration.
```javascript
var keywordValidator;
```

Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Actually, given the layout of this file, we should probably be declaring each of these vars on the same line where we assign them. Assigning a variable far away from its declaration (170 lines away here) often leads to bugs/typos/omissions.

Naturally, let's not change it for all of the other variables, but what about moving this declaration below:
```javascript
var keywordValidator = (function () {
    // ...
}());
```

File: cms/static/js/views/utils/view_utils.js (hunk 16 above)
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Optionally, you could combine these two lines:
```javascript
var keywordsFound = string.match(regexp) || [];
```

File: cms/static/js/views/utils/view_utils.js [hunk 25]
```diff
@@ -173,6 +174,33 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
             return false;
         };
 
+        keywordValidator = (function () {
+            var regexp = /%%+[^%]+%%/g;
+            var keywords = ['%%USER_ID%%', '%%USER_FULLNAME%%', '%%COURSE_DISPLAY_NAME%%', '%%COURSE_END_DATE%%'];
+            var validate = function (string) {
+                var regex_match = string.match(regexp);
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: JavaScript convention expects variables to be camelCased, not snake_cased.

File: cms/static/js/views/settings/main.js [hunk 26]
```diff
@@ -367,12 +368,18 @@ var DetailsView = ValidatingView.extend({
             message = this.post_enrollment_email_elem.val();
         }
 
-        $.post("/settings/send_test_enrollment_email",
+        validation = ViewUtils.keywordValidator.validate_string(message);
```
Comment (stvstnfrd) [inline cms/static/js/views/settings/main.js]: See note below about camelCase vs snake_case.

**Comment by stvstnfrd** [comment]
Could you add a stanza to the commit message describing how to manually test/use this feature?

File: cms/static/js/views/settings/main.js (hunk 11 above)
Comment (kluo) [inline cms/static/js/views/settings/main.js]: They are contained in the endpoint now

File: cms/static/js/views/utils/view_utils.js (hunk 24 above)
Comment (kluo) [inline cms/static/js/views/utils/view_utils.js]: Sure, but doesn't this go against hoisting?

File: cms/static/js/views/utils/view_utils.js (hunk 23 above)
Comment (kluo) [inline cms/static/js/views/utils/view_utils.js]: Good question / catch, I took this code verbatim from Nick's course announcement branch :P I'll fix that.

File: cms/static/js/views/utils/view_utils.js (hunk 24 above)
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Ah yeah, I guess I sort of see this as an exception to the rule. The way
this module is constructed, `keywordValidator` and related vars are only
really used for a single assignment. So if this were applied throughout
(I'm not proposing to change them now), the variables could/should be
declared on the same line on which the assignment occurs.

```javascript
var hasChangedAttributes = function (model, attributes) {
};

var keywordValidator = (function () {
}());
```

The final call is up to you, though.

Comment (kluo) [inline cms/static/js/views/utils/view_utils.js]: Ok I'll go ahead and declare it where it's assigned then, cleaner diff too

File: cms/startup.py [hunk 27]
```diff
@@ -9,6 +9,8 @@
 
 from django_startup import autostartup
 from monkey_patch import django_utils_translation
+from util import keyword_substitution
+from lms.startup import get_keyword_function_map
```
Comment (stvstnfrd) [inline cms/startup.py]: I know that importing from CMS/LMS into common is forbidden. 

What's the policy w/r/t importing from LMS->CMS (and vice-versa)?

Ideally, the `get_keyword_function_map` function should be stored in a `common` location. But, I'm assuming that this method already exists upstream? Assuming so, let's defer and not diverge more than necessary.

File: cms/startup.py (hunk 2 above)
Comment (stvstnfrd) [inline cms/startup.py]: A future enhancement could be to port this logic over to my fancy `common/djangoapps/monkey_patch` library ;)

File: cms/static/js/views/utils/view_utils.js [hunk 28]
```diff
@@ -173,6 +173,31 @@ define(["jquery", "underscore", "gettext", "js/views/feedback_notification", "js
...
+                        return undefined;
+                    }
+                });
+
+                return {
+                    'isValid': keywordsInvalid.length === 0,
+                    'keywordsInvalid': keywordsInvalid
+                }
+
+            };
```
Comment (stvstnfrd) [inline cms/static/js/views/utils/view_utils.js]: Remove the trailing `;`; it's superfluous for named functions.

**Comment by stvstnfrd** [comment]
The only outstanding actionable item is to remove the superfluous `;`, after that, +1.

File: cms/startup.py (hunk 27 above)
Comment (jbau) [inline cms/startup.py]: I think the cms settings files import lms:
https://github.com/Stanford-Online/edx-platform/blob/master/cms/envs/common.py#L30

One could argue that startup.py is also such a special case instance, where importing like we do is ok.

Comment (stvstnfrd) [inline cms/startup.py]: Yeah, I'm cool with it; more curious about convention than anything, so thanks for the link.
